File: ruby/basic-attendant.rb
# ===========
# Tropo basic auto attendant sample application in Ruby
# v0.9 Jason Goecke - Modeled on JS example by jrt
# ===========

# define the list of contacts

Contacts = { :jsgoecke => { :name_choices => "Jason, Jason Goecke", :number => "18005551212" },
             :jicksta  => { :name_choices => "Jay, Jay Phillips",   :number => "14155551212" } }

# turn the contacts into a comma seperated list of names
def list_names
  list_of_names = String.new
  Contacts.each { |contact| list_of_names = list_of_names + contact[1][:name_choices] + ", " }
  return list_of_names
end

# turn the options into something like:
# jsgoecke(Jason, Jason Goecke),jicksta(Jay, Jay Phillips)
def list_options
  list_of_options = String.new
  Contacts.each { |contact| list_of_options = list_of_options + contact[0].to_s + "(" + contact[1][:name_choices] + ")," }
  return list_of_options  
end

# start

#answer the phone and play the initial greeting

answer 30
prompt "hello, and thank you for calling."

options = { "repeat"  => 3,
            "timeout" => 5,
            "choices" => list_options,
            # Create a callback on an event
            "onEvent" => lambda { |event|
              case event.name
              # When the user has selected a valid option for the list provided in the prompt
              when "choice"
                # Tell the user that we understand them and that we will transfer them
                say "ok, you said " + event.value + ".  Please hold while I transfer you."
                options = { "answerOnMedia" => false,
                            "callerID"      => "tel:+14076179024",
                            "timeout"       => 60.3456,
                            "method"        => "bridged", # fixed to bridged currently
                            "playrepeat"    => 3,
                            "playvalue"     => "Ring... Ring... Ring...",
                            "choices"       => "1,2,3,4,5,6,7,8,9,0,*,#",
                            # Create a series of callbacks to reflect the status of the call result
                            "onSuccess"     => lambda { |event| log "Xfer Success!: " + event.inspect.to_s },
                            "onError"       => lambda { |event| log "Xfer Error!: " + event.inspect.to_s },
                            "onTimeout"     => lambda { |event| log "Xfer Timeout!: " + event.inspect.to_s },
                            "onCallFailure" => lambda { |event| log "Xfer Failure!: " + event.inspect.to_s },
                            "onChoice"      => lambda { |event| log "Xfer Cancelled!: " + event.inspect.to_s }
                          }
                transfer_event = transfer "sip:9" + Contacts[event.value.to_sym][:number] + "@10.6.63.201", options
              # When a user has said something not understood to be on the list provided in the prompt
              when "badChoice"
                say "I'm sorry, I didn't understand what you said."
              # When the user has remained silent up to the timeout
              when "timeout"
                say "I'm sorry, I didn't hear anything."
              end
            }
          }

# prompt the user for the name of the person they desire          
event = prompt "Who would you like to call?  Just say " + list_names, options

# Hangup the call once complete
hangup

File: php/basic-attendant.php
 <?php
/*
 * Copyright (c) 2009 - 2015 Tropo, now part of Cisco
 * Released under the MIT license. See the file LICENSE
 * for the complete license
 */


// ===========
// Tropo basic auto attendant sample application
// v1 jrt -> php rsc
// ===========


// -----------
// turn the contacts into a comma seperated list of names

function listNames( $theContacts )
{
  $s = '';

  foreach( $theContacts as $index => $contact ) {
   if ($s != '') $s .= ", " ;
    $s .= $index;
  }

  return $s;
}  //function listNames

// -----------
// turn the contacts into a comma seperated list of options for each contact (simple grammar)

function listOptions( $theContacts )
{
  $s ='';

  foreach( $theContacts as $index => $contact ) {
    if ($s != '') $s .= ", " ;
    $s .= $index . " (" . $contact['nameChoices'] . ")";
  }

  return $s;
}

// -----------
// start
// -----------

// define the list of contacts

$contacts = array("nicole"=> array("nameChoices" => "Nicole, Nicole Williams", "number" => "12067927251" ),
                  "april" => array("nameChoices" => "April, April Smith",      "number" => "14129272367" ),
                  "gina"  => array("nameChoices" => "Gina, Gina Anderson",     "number" => "17135744782" ) );


// answer the phone and play the initial greeting

answer( 30 );

say( "hello, and thank you for calling." );

// prompt the user for the name of the person they desire
$event = ask( "Who would you like to call?  Just say " . listNames( $contacts ),
            array (
              "repeat"  => "3",
              "timeout" => "7",
              "choices" => listOptions( $contacts ),
              "onTimeout" => create_function( '$event', 'say( "I\'m sorry, I didn\'t hear anything." );'),
              "onBadChoice" => create_function( '$event', 'say( "I\'m sorry, I didn\'t understand what you said." );')
            ) 
         );

// if they made a choice, transfer to that person

if ($event->name == 'choice')
{
  say( "ok, you said " . $event->value . ".  Please hold while I transfer you." );

$ne = transfer( "tel:+" . $contacts[ $event->value ]['number'],
         array (
         "answerOnMedia" => false,
         "callerID"      => "14074181800",
         "timeout"       => 60.3456,
         "playrepeat"    => 3,
         "playvalue"     => "Ring... Ring... Ring...",
         "choices"       => "1,2,3,4,5,6,7,8,9,0,*,#",
         "onSuccess"     => create_function('$event', '_log("*********transfered to:  " . $event->value->calledId);'),  //~!~ fix
         "onError"       => create_function('$event', '_log("*********transfer error");'),
         "onTimeout"     => create_function('$event', '_log("*********transfer timeout");'),
         "onCallFailure" => create_function('$event', '_log("*********transfer failed");'),
         "onChoice"      => create_function('$event', '_log("*********transfer canceled");')
         ) 
      );

  _log( "transfer event.name  = " . $ne->name );
  _log( "transfer event.value = " . $ne->value );
  say( "Goodbye" );

}

hangup();
?>